lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp

```diff
@@ -192,185 +217,354 @@ class AccessConflictAnalysis {
       return const_cast<Result &>(*this).getAccessInfo(beginAccess);
     }
   };
+  using AccessedStorageLocs = llvm::SmallDenseSet<AccessedStorage, 8>;
 
 private:
-  SILFunction *F;
-  PostOrderFunctionInfo *PO;
+  LoopRegionFunctionInfo *LRFI;
   AccessedStorageAnalysis *ASA;
 
-  /// Tracks the in-scope accesses at the end of each block, for the purpose of
-  /// finding nested conflicts.  (Out-of-scope accesses are currently only
-  /// tracked locally for the purpose of merging access scopes.)
-  llvm::SmallDenseMap<SILBasicBlock *, DenseAccessSet, 32> blockOutAccess;
+  using RegionToRegionStorageMap =
+      llvm::SmallDenseMap<unsigned, RegionStorage *>;
+  RegionToRegionStorageMap regionToStorageMap;
+  AccessedStorageLocs storageAccesses;
 
   Result result;
-
 public:
-  AccessConflictAnalysis(SILFunction *F, PostOrderFunctionInfo *PO,
+  AccessConflictAnalysis(LoopRegionFunctionInfo *LRFI,
                          AccessedStorageAnalysis *ASA)
-      : F(F), PO(PO), ASA(ASA) {}
+      : LRFI(LRFI), ASA(ASA) {}
+  ~AccessConflictAnalysis();
 
   Result &&analyze() &&;
-
 protected:
   void identifyBeginAccesses();
+  void identifyStorageLocations();
+  void processBlockRegion(unsigned id);
+  void processLoopRegion(unsigned id);
 
-  void visitBeginAccess(BeginAccessInst *beginAccess,
-                        SparseAccessSet &accessMap);
+  void mergeBBPredAccesses(unsigned succID);
 
-  void visitEndAccess(EndAccessInst *endAccess, SparseAccessSet &accessMap);
+  void visitBeginAccess(BeginAccessInst *beginAccess);
+  void visitEndAccess(EndAccessInst *endAccess);
+  void detectApplyConflicts(const DenseAccessSet &conflictFreeSet,
+                            RegionStorage *applyStorage,
+                            const FunctionAccessedStorage &callSiteAccesses,
+                            const FullApplySite &fullApply);
 
-  void visitFullApply(FullApplySite fullApply, SparseAccessSet &accessMap);
+  void visitFullApply(FullApplySite fullApply);
 
-  void mergePredAccesses(SILBasicBlock *succBB,
-                         SparseAccessSet &mergedAccesses);
+  void calcBottomUpOrder(LoopRegion *region,
+                         llvm::SmallVectorImpl<unsigned> &worklist);
 
-  void visitBlock(SILBasicBlock *BB);
+  RegionStorage *getRegionStorageFromIns(SILInstruction *instr);
 };
 
-/// A sparse set of in-flight accesses.
-///
-/// Explodes a DenseAccessSet into a temporary sparse set for comparison
-/// and membership.
-///
-/// The AccessConflictAnalysis result provides the instruction to index
-/// mapping.
-class SparseAccessSet {
+/// The per-region result of AccessConflictAnalysis.
+class RegionStorage {
```

Comment:
The name for 'RegionStorage' is really bad ( I assume 'Storage' here means to store the region data), because it can easily be confused with access storage. I would call it 'RegionInfo' or something like this.